cardinal/world.go

```diff
@@ -357,104 +448,325 @@ func (w *World) Shutdown() error {
+// any calls to WaitForNextTick that happen after a shutdown will not block.
+func (w *World) drainChannelsWaitingForNextTick() {
+	go func() {
+		for ch := range w.addChannelWaitingForNextTick {
+			close(ch)
+		}
+	}()
+}
+
+func (w *World) drainEndLoopChannels() {
+	go func() {
+		for range w.endGameLoopCh { //nolint:revive // This pattern drains the channel until closed
+		}
+	}()
+}
+
+// AddTransaction adds a transaction to the transaction queue. This should not be used directly.
+// Instead, use a MessageType.AddToQueue to ensure type consistency. Returns the tick this transaction will be
+// executed in.
+func (w *World) AddTransaction(id types.MessageID, v any, sig *sign.Transaction) (
+	tick uint64, txHash types.TxHash,
+) {
+	// TODO: There's no locking between getting the tick and adding the transaction, so there's no guarantee that this
+	// transaction is actually added to the returned tick.
+	tick = w.CurrentTick()
+	txHash = w.txQueue.AddTransaction(id, v, sig)
+	return tick, txHash
+}
+
+func (w *World) AddEVMTransaction(
+	id types.MessageID,
+	v any,
+	sig *sign.Transaction,
+	evmTxHash string,
+) (
+	tick uint64, txHash types.TxHash,
+) {
+	tick = w.CurrentTick()
+	txHash = w.txQueue.AddEVMTransaction(id, v, sig, evmTxHash)
+	return tick, txHash
+}
+
+func (w *World) UseNonce(signerAddress string, nonce uint64) error {
+	return w.redisStorage.UseNonce(signerAddress, nonce)
+}
+
+func (w *World) loadGameState() error {
+	if w.WorldState != WorldStateInit {
+		return eris.New("cannot load game state multiple times")
+	}
+
+	// TODO(scott): footgun. so confusing.
+	if err := w.entityStore.RegisterComponents(w.registeredComponents); err != nil {
+		closeErr := w.entityStore.Close()
+		if closeErr != nil {
+			return eris.Wrap(err, closeErr.Error())
+		}
+		return err
+	}
+
+	recoveredTxs, err := w.recoverGameState()
 	if err != nil {
 		return err
 	}
+
+	// Engine is now ready to run
+	w.WorldState = WorldStateReady
+
+	// Recover the last tick, not to be confused with RecoverFromChain
+	// It's ambigious, but screw it for now.
+	if recoveredTxs != nil {
+		w.txQueue = recoveredTxs
+		if err = w.Tick(context.Background()); err != nil {
+			return err
+		}
+	}
+	w.receiptHistory.SetTick(w.CurrentTick())
+
 	return nil
 }
 
-func (w *World) Engine() *ecs.Engine {
-	return w.instance
+func (w *World) Namespace() Namespace {
+	return w.namespace
 }
 
-func (w *World) CurrentTick() uint64 {
-	return w.instance.CurrentTick()
+func (w *World) GetQueryByName(name string) (engine.Query, error) {
+	if q, ok := w.nameToQuery[name]; ok {
+		return q, nil
+	}
+	return nil, eris.Errorf("query with name %s not found", name)
 }
 
-func (w *World) Tick(ctx context.Context) error {
-	return w.instance.Tick(ctx)
+func (w *World) GetMessageByName(name string) (types.Message, bool) {
+	for _, msg := range w.msgManager.GetRegisteredMessages() {
+		if msg.Name() == name {
+			return msg, true
+		}
+	}
+	return nil, false
```

Comment:
feels like this logic should be moved to the manager. the messages are already in a map over there, so should be a very smol method